### PR TITLE
Login: Fix popup message to state account was blocked

### LINKED ISSUE #43531: Login: Grafana Login Popup Message Does Not State Account Was Blocked
**What happened**:
Failed login on Grafana, popup message saying username or password is incorrect.

**What you expected to happen**:
In the logs, the server shows that the account was blocked, but the login message still shows the same error message.

**How to reproduce it (as minimally and precisely as possible)**:
Fail a login 5 times.

**Anything else we need to know?**:
The popup should show that the account was blocked, just as the logs show the account was blocked.

**Environment**:
- Grafana version: 8.3.3

### PR DESCRIPTION
**Pull request by axelavargas**
**What this PR does / why we need it**:

The popup message saying username or password is incorrect was being sent even when the account of the user was blocked. 

I added some conditions to make it more specific.

Before
BeforeErrorNotCorrect

After
AfterRightMessage

**Which issue(s) this PR fixes**:

Fixes #43531

**Special notes for your reviewer**:

I have no prior experience in Go, this is just an attempt to see if good enough 😅

### CODE REVIEW

**Comment by joshhunt** [comment]
Might be good for @grafana/docs-squad to pass their eyes over this?

**Comment by hugohaggmark** [comment]
Have we tried to replicate this without the `app_mode = development` in the ini-file? I know that responses from the backend api differ depending on that setting in the ini-file.
https://grafana.com/docs/grafana/latest/administration/configuration/#app_mode

**Comment by hugohaggmark** [comment]
> > Have we tried to replicate this without the `app_mode = development` in the ini-file? I know that responses from the backend api differ depending on that setting in the ini-file. [grafana.com/docs/grafana/latest/administration/configuration/#app_mode](https://grafana.com/docs/grafana/latest/administration/configuration/#app_mode)
> 
> Hey @hugohaggmark 😄 , I did the test on the main branch without the `app_mode = development `, I encounter the same error message " invalid username or password", after I did the same test using my fix branch and the error message is consistent (a better error message)🚀

Did you change in .bra.toml file as well from dev to production? (if you run bra run or make run)

**Comment by axelavargas** [comment]
> Did you change in .bra.toml file as well from dev to production? (if you run bra run or make run)

Nope, I didn't know the existence of bra.toml 😅 , testing it now

**Comment by axelavargas** [comment]
> > Did you change in .bra.toml file as well from dev to production? (if you run bra run or make run)
> 
> Nope, I didn't know the existence of bra.toml sweat_smile , testing it now

Alright, I did the test and got the same result as before, with my fix branch I have the right message :)

**Comment by hugohaggmark** [comment]
Sorry I'm a bit tired, so I just want to make sure, but when I change my `.bra.toml` file and `custom.ini` to use production and restart `make run` this is what I get running your branch:
image

**Comment by axelavargas** [comment]
> Sorry I'm a bit tired, so I just want to make sure, but when I change my `.bra.toml` file and `custom.ini` to use production and restart `make run` this is what I get running your branch: image

No worries Hugo 🤗, just to double-check you are getting this screen after trying more than 5 times with the wrong password?

**Comment by hugohaggmark** [comment]
> > Sorry I'm a bit tired, so I just want to make sure, but when I change my `.bra.toml` file and `custom.ini` to use production and restart `make run` this is what I get running your branch: image
> 
> No worries Hugo 🤗, just to double-check are you getting this screen after trying more than 5 times with the wrong password?

Yes

**Comment by hugohaggmark** [comment]
My bad, I didn't try enough times in a row 🙈

File: pkg/api/login.go [hunk 1]
```diff
@@ -210,7 +210,21 @@ func (hs *HTTPServer) LoginPost(c *models.ReqContext) response.Response {
 	err := bus.Dispatch(c.Req.Context(), authQuery)
 	authModule = authQuery.AuthModule
 	if err != nil {
-		resp = response.Error(401, "Invalid username or password", err)
+		var msg string
+
+		if errors.Is(err, login.ErrTooManyLoginAttempts) {
+			msg = "Too many consecutive incorrect login attempts for user - login for user temporarily blocked"
```
Comment (papagian) [inline pkg/api/login.go:216]: alternatively, you could do the following in order to use [this text](https://github.com/grafana/grafana/blob/77347a37b1741a592c46cf5ff7cd5f8d812795b3/pkg/login/auth.go#L18).
```suggestion
			msg = err.Error()
```

You can also apply the same for the other cases.

File: pkg/api/login.go [hunk 2]
```diff
@@ -210,7 +210,21 @@ func (hs *HTTPServer) LoginPost(c *models.ReqContext) response.Response {
...
+
+		if errors.Is(err, login.ErrInvalidCredentials) {
+			msg = "Invalid username or password"
+		}
+
+		if errors.Is(err, models.ErrUserNotFound) {
+			msg = "User not found"
+		}
+
+		resp = response.Error(401, msg, err)
```
Comment (papagian) [inline pkg/api/login.go:227]: If the error is not any of the above cases then the message will be empty (the default string value in Go). So you need [initialise it](https://github.com/grafana/grafana/blob/547310ed1b813393ed0ae6d72789ece084bc4a4c/pkg/api/login.go#L213) to a default like `Invalid username or password`.

File: pkg/api/login.go [hunk 3]
```diff
@@ -210,7 +210,21 @@ func (hs *HTTPServer) LoginPost(c *models.ReqContext) response.Response {
...
+			msg = "Too many consecutive incorrect login attempts for user - login for user temporarily blocked"
+		}
+
+		if errors.Is(err, login.ErrInvalidCredentials) {
+			msg = "Invalid username or password"
+		}
+
+		if errors.Is(err, models.ErrUserNotFound) {
+			msg = "User not found"
+		}
```
Comment (sakjur) [inline pkg/api/login.go:225]: Don't tell a potentially hostile user that the username isn't valid.
 https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Authentication_Cheat_Sheet.md#authentication-and-error-messages

**Review by sakjur — CHANGES_REQUESTED** [review]

File: pkg/api/login.go [hunk 4]
```diff
@@ -210,7 +210,14 @@ func (hs *HTTPServer) LoginPost(c *models.ReqContext) response.Response {
 	err := bus.Dispatch(c.Req.Context(), authQuery)
 	authModule = authQuery.AuthModule
 	if err != nil {
-		resp = response.Error(401, "Invalid username or password", err)
+		var msg = "Invalid username or password"
+
+		if errors.Is(err, login.ErrTooManyLoginAttempts) {
+			msg = err.Error()
```
Comment (sakjur) [inline pkg/api/login.go:216]: You won't get a capital first letter now, if that matters to you 🙂

**Review by sakjur — DISMISSED** [review]
:+1: Now it looks good

**Comment by sakjur** [comment]
> I made some changes, but I am a bit unsure now about the whole approach, should we provide the user information about the account being locked?

That is a very good question.

It depends a little imo.

Let's say we're attacking Grafana, trying to figure out the password to the user `emil`, and failing N times, locking the user.
If we'd now start seeing "try again later" instead of "invalid credentials", does that tell us 1. that the user exists 2. if we've guessed the right password?

What we'd want, to prevent disclosure, is that the brute force protection is applied in such a way that you wouldn't gain any more information about either of those two (i.e. even guessing a user that doesn't exist should trigger the brute force protection and the brute force protection should be applied before trying to log the user in). I think, if we can make sure that's true, that this is an improvement worth making 🙂

**Comment by axelavargas** [comment]
> > I made some changes, but I am a bit unsure now about the whole approach, should we provide the user information about the account being locked?
> 
> That is a very good question.
> 
> It depends a little imo.
> 
> Let's say we're attacking Grafana, trying to figure out the password to the user `emil`, and failing N times, locking the user. If we'd now start seeing "try again later" instead of "invalid credentials", does that tell us 1. that the user exists 2. if we've guessed the right password?
> 
> What we'd want, to prevent disclosure, is that the brute force protection is applied in such a way that you wouldn't gain any more information about either of those two (i.e. even guessing a user that doesn't exist should trigger the brute force protection and the brute force protection should be applied before trying to log the user in). I think, if we can make sure that's true, that this is an improvement worth making slightly_smiling_face

Hey @sakjur 👋🏾 , 🤔 I believe I am lacking important context here, do we have in place brute force protection? so far I only noticed that after x amount of attempts the user is temporarily blocked, but I am not sure if this blocked is based on time? or if we are blocking the  IP, sending emails?  I am leaning towards closing my PR and sending the relevant issue to the backend squad, it is going to be easier to make those decisions with the right context.  

Let me know what you think 😄

**Comment by sakjur** [comment]
@axelavargas I think you made a really good call to close this for now 🙂

I agree with you that context is key here, the difference between a security vulnerability and a great UX feature relies on figuring out things about other parts of the system (like you're saying)— and if that's not something you have time for at the moment, it seems like the right thing to do to not do anything at all.

Still, I hope someone picks this up, the base is reasonable after we've made sure that the error messages doesn't let you enumerate users or passwords 😄

**Pull request closed without merging**